### PR TITLE
feat: add backoff to BulkWriter

### PR DESCRIPTION
**Pull request by thebrianchen**
Combination of [1](https://github.com/googleapis/nodejs-firestore/pull/1447) and [2](https://github.com/googleapis/nodejs-firestore/pull/1451) from node.

### CODE REVIEW

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java [hunk 1]
```diff
@@ -1097,6 +1136,159 @@ public void failsWritesAfterAllRetryAttemptsFail() throws Exception {
...
 
+  @Test
+  public void appliesMaxBackoffOnRetriesForResourceExhausted() throws Exception {
+    final int[] retryAttempts = {0};
+    final ScheduledExecutorService timeoutExecutor =
+        new ScheduledThreadPoolExecutor(1) {
+          @Override
+          @Nonnull
+          public ScheduledFuture<?> schedule(Runnable command, long delay, TimeUnit unit) {
+            int expected =
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java:1147]: This seems unused.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java [hunk 2]
```diff
@@ -1097,6 +1136,159 @@ public void failsWritesAfterAllRetryAttemptsFail() throws Exception {
...
+
+    bulkWriter.addWriteErrorListener(
+        new WriteErrorCallback() {
+          public boolean onError(BulkWriterException error) {
+            return error.getFailedAttempts() < 5;
+          }
+        });
+    bulkWriter.create(doc1, LocalFirestoreHelper.SINGLE_FIELD_MAP);
+    bulkWriter.flush();
+    bulkWriter.set(doc1, LocalFirestoreHelper.SINGLE_FIELD_MAP);
```
Comment (schmidt-sebastian) [inline google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java:1288]: Should this be a different doc?

**Review by schmidt-sebastian — COMMENTED** [review]
Some questions about the tests, but implementation LGTM.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java (hunk 2 above)
Comment (thebrianchen) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java]: changed.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java [hunk 3]
```diff
@@ -1097,6 +1136,159 @@ public void failsWritesAfterAllRetryAttemptsFail() throws Exception {
...
+          (BulkWriterOperation.DEFAULT_BACKOFF_INITIAL_DELAY_MS
+              * BulkWriterOperation.DEFAULT_BACKOFF_FACTOR)
+    };
+    final int[] retryAttempts = {0};
+    final ScheduledExecutorService timeoutExecutor =
+        new ScheduledThreadPoolExecutor(1) {
+          @Override
+          @Nonnull
+          public ScheduledFuture<?> schedule(Runnable command, long delay, TimeUnit unit) {
+            if (delay > 0) {
```
[the remark this answers is not in the document]
Comment (thebrianchen) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java:1216]: This test increments `retryAttempts` so it's ok.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriterOperation.java [hunk 4]
```diff
@@ -106,6 +128,16 @@ public void onSuccess(Boolean shouldRetry) {
     return callbackFuture;
   }
 
+  private void updateBackoffDuration() {
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriterOperation.java:131]: I missed the `*`.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java [hunk 5]
```diff
@@ -1097,6 +1136,159 @@ public void failsWritesAfterAllRetryAttemptsFail() throws Exception {
...
+    final ScheduledExecutorService timeoutExecutor =
+        new ScheduledThreadPoolExecutor(1) {
+          @Override
+          @Nonnull
+          public ScheduledFuture<?> schedule(Runnable command, long delay, TimeUnit unit) {
+            int expected =
+                (int)
+                    (BulkWriterOperation.DEFAULT_BACKOFF_INITIAL_DELAY_MS
+                        * Math.pow(1.5, retryAttempts[0]));
+            if (delay > 0) {
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java:1152]: That's a good way to address this.

**Review by schmidt-sebastian — APPROVED** [review]

**Pull request merged**